File: Cracking The C.I./cap05/Q5_1.cpp
#include <iostream>
#include <bitset>
using namespace std;

int combineNumber(bitset<32> _N, bitset<32> _M, int j, int i) {
	int N = _N.to_ulong();
	int M = _M.to_ulong();
	cout << "N:     " << _N << endl;
	cout << "M:     " << _M << endl;

	// clear bits from j =< x <= i
	int allOnes = ~0;
	int leftPart = allOnes << (j+1);
	cout << "Left:  " << bitset<32>(leftPart) << endl;
	int rightPart = (1 << i) - 1;
	cout << "roght: " << bitset<32>(rightPart) << endl;
	int mask = leftPart | rightPart;
	cout << "mask:  " << bitset<32>(mask) << endl;

	N = N & mask;

	M = M << i;
	N = N | M;
	return N;
}

int main() {
	int result = combineNumber(
		bitset<32>("100000000000"), 
		bitset<32>("10011"),
		6,
		2);
	bitset<32> realResult(result);
	cout << "result:"<<realResult << endl;
	return 0;
};

File: Cracking The C.I./cap05/Q5_1.java
import java.util.*;

public class Q5_1 {

	public static int updateBits(String _N, String _M, int j, int i) {
		int N = Integer.parseInt(_N, 2);
		int M = Integer.parseInt(_M, 2);

		System.out.println("N:  "+Integer.toBinaryString(N));
		System.out.println("M:  "+Integer.toBinaryString(M));

		int allOnes = ~0;
		int leftPart = allOnes << (j+1);
		int rightPart = (1 << i) - 1;

		int mask = leftPart | rightPart;
		System.out.println("mask:  "+Integer.toBinaryString(mask));

		N = N & mask;

		M = M << i;
		N = N | M;
		return N;
	}

	public static void main(String[] args) {
		int result = updateBits(
			"100000000000",
			"10011",
			6,
			2);
		System.out.println("result: "+Integer.toBinaryString(result));
	}

}